File: UnidadesExternas_vb/USBUtil.vb
'--------------------------------------------------------------------------------
' USBUtil                                                       (11/abr/22 16.32)
' Versión para Visual Basic.                                    (11/abr/22 20.42)
' Utilidades para saber las unidades externas (extraíbles o fijas)
'
' Versión para .NET 5.0
'
' (c) Guillermo Som (Guille), 2022
'--------------------------------------------------------------------------------
Option Strict On
Option Infer On

Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
' Hay que instalar System.Management de NuGet. versión 5.0 poara .NET Framework o .NET 5.0 o inferior.
Imports System.Management

Public Class USBUtil

    ''' <summary>
    ''' Comprueba si es una unidad externa.
    ''' </summary>
    ''' <param name="elPath">Path completo del que se extraerá la letra de unidad.</param>
    ''' <returns></returns>
    Public Shared Function EsUnidadExterna(ByVal elPath As String) As Boolean
        Dim disco = Path.GetPathRoot(Path.GetFullPath(elPath))
        Dim usbD = GetUsbDriveLetters()
        Return usbD.Contains(disco)
    End Function

    ''' <summary>
    ''' Las letras de las unidades externas conectadas por USB.
    ''' </summary>
    ''' <returns></returns>
    Public Shared Function GetUsbDriveLetters() As List(Of String)
        ' Hay que usar External para que también tenga en cuenta los USB de tipo fijo.
        Dim usbDrivesLetters = From drive In New ManagementObjectSearcher("select * from Win32_DiskDrive WHERE MediaType like '%External%' OR InterfaceType='USB'").[Get]().Cast(Of ManagementObject)()
                               From o In drive.GetRelated("Win32_DiskPartition").Cast(Of ManagementObject)()
                               From i In o.GetRelated("Win32_LogicalDisk").Cast(Of ManagementObject)()
                               Select String.Format("{0}\", i("Name"))
        Return usbDrivesLetters.ToList()
    End Function

End Class

File: UnidadesExternas_cs/USBUtil.cs
//--------------------------------------------------------------------------------
// USBUtil                                                      (11/abr/22 16.32)
// Utilidades para saber las unidades externas (extraíbles o fijas)
//
// Versión para .NET 5.0
//
// (c) Guillermo Som (Guille), 2022
//--------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
// Hay que instalar System.Management de NuGet. versión 5.0 poara .NET Framework o .NET 5.0 o inferior.
using System.Management;


namespace UnidadesExternas_cs
{
    public class USBUtil
    {
        /// <summary>
        /// Comprueba si es una unidad externa.
        /// </summary>
        /// <param name="elPath">Path completo del que se extraerá la letra de unidad.</param>
        /// <returns></returns>
        public static bool EsUnidadExterna(string elPath)
        {
            var disco = Path.GetPathRoot(Path.GetFullPath(elPath));
            var usbD = GetUsbDriveLetters();
            return usbD.Contains(disco);
        }

        // Código basado en dos ejemplos de "la red".
        // https://stackoverflow.com/a/31560283/14338047
        // https://stackoverflow.com/a/10018438/14338047

        /// <summary>
        /// Las letras de las unidades externas conectadas por USB.
        /// </summary>
        /// <returns></returns>
        public static List<string> GetUsbDriveLetters()
        {
            // Hay que usar External para que también tenga en cuenta los USB de tipo fijo.
            var usbDrivesLetters = from drive in new ManagementObjectSearcher("select * from Win32_DiskDrive WHERE MediaType like '%External%' OR InterfaceType='USB'").Get().Cast<ManagementObject>()
                                   from o in drive.GetRelated("Win32_DiskPartition").Cast<ManagementObject>()
                                   from i in o.GetRelated("Win32_LogicalDisk").Cast<ManagementObject>()
                                   select string.Format("{0}\\", i["Name"]);

            return usbDrivesLetters.ToList();
        }

    }
}
